File: ipification-sdk-android-java/app/src/main/java/com/ipification/demoapp/manager/IMHelper.java
package com.ipification.demoapp.manager;

import android.util.Log;

import com.ipification.demoapp.Urls;
import com.ipification.demoapp.callback.TokenCallback;
import com.ipification.mobile.sdk.android.model.IMSession;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class IMHelper {

    private static final String TAG = "IMHelper";
    public static String currentState = null;
    public static String currentToken = null;
    private static MediaType JSON_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");
    public static IMSession sessionInfo;

    public static void registerDevice(String deviceToken, String state) {
        if (deviceToken.isEmpty() || state == null || state.equals("")) {
            Log.d(TAG, "deviceToken or state null. failed");
            return;
        }
        MediaType JSON = MediaType.parse("application/json; charset=utf-8");
        String url = Urls.DEVICE_TOKEN_REGISTRATION_URL;
        String json = "{\"device_id\": \"" + state + "\",\"device_token\": \"" + deviceToken + "\", \"device_type\":\"android\"}";
        Log.d(TAG, "registerDevice " + json);

        OkHttpClient client = new OkHttpClient.Builder().build();
        Request request = new Request.Builder().url(url).post(RequestBody.create(JSON, json))
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                if (response.isSuccessful()) {
                    Log.d(TAG, "registered successfully - " + response.body().string());
                } else {
                    Log.e(TAG, "registered failed " + response.body().string());
                }
            }

            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.e(TAG, "registered failed" + e.getMessage());
            }
        });
    }

    public static void signIn(String state, TokenCallback callback) {
        String url = Urls.AUTOMODE_SIGN_IN_URL;
        JSONObject jsonObject = new JSONObject();
        try {
            jsonObject.put("state", state);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        RequestBody body = RequestBody.create(jsonObject.toString(), JSON_MEDIA_TYPE);
        Request request = new Request.Builder()
                .url(url)
                .post(body)
                .build();

        OkHttpClient client = new OkHttpClient.Builder().build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                String error = e.getLocalizedMessage();
                Log.e(TAG, "SignIn onFailure: " + error, e);
                callback.onError(error != null ? error : "postUserInfo - onFailure");
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                try {
                    String responseBody = response.body().string();

                    if (response.isSuccessful()) {
                        callback.onSuccess(responseBody != null ? responseBody : "");
                    } else {
                        callback.onError(responseBody != null ? responseBody : "");
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error processing response", e);
                    callback.onError(e.getLocalizedMessage() != null ? e.getLocalizedMessage() : "Error processing response");
                }
            }
        });
    }

}

File: ipification-sdk-android-kotlin/app/src/main/java/com/ipification/demoapp/manager/IMHelper.kt
package com.ipification.demoapp.manager

import android.util.Log
import com.ipification.demoapp.Urls
import com.ipification.demoapp.callback.TokenCallback
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
class IMHelper {
    companion object Factory {
        private const val TAG = "IMHelper"
        var currentState: String? = null
        internal var deviceToken: String? = null
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaTypeOrNull()

        fun registerDevice(dToken: String? = deviceToken, state: String? = currentState) {
            if (dToken.isNullOrEmpty() || state.isNullOrEmpty()) {
                Log.d(TAG, "deviceToken or state null. failed")
                return
            }

            val url = Urls.DEVICE_TOKEN_REGISTRATION_URL
            val json = "{\"device_id\":\"$currentState\",\"device_token\":\"${deviceToken}\", \"device_type\":\"android\"}"
            Log.d(TAG, "registerDevice $json")

            val requestBody = json.toRequestBody(JSON_MEDIA_TYPE)
            val request = Request.Builder().url(url).post(requestBody).build()

            val client = OkHttpClient.Builder().build()
            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    Log.e(TAG, "Registration failed", e)
                }

                override fun onResponse(call: Call, response: Response) {
                    Log.d(TAG, "Registration successful - ${response.body?.string()}")
                }
            })
        }

        fun signIn(state: String, callback: TokenCallback) {
            val url = Urls.AUTOMODE_SIGN_IN_URL
            val jsonObject = JSONObject().apply {
                put("state", state)
            }

            val body = jsonObject.toString().toRequestBody(JSON_MEDIA_TYPE)
            val request = Request.Builder().url(url).post(body).build()

            val client = OkHttpClient.Builder().build()
            client.newCall(request).enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    val error = e.localizedMessage
                    Log.e(TAG, "SignIn onFailure: $error", e)
                    callback.onError(error ?: "postUserInfo - onFailure")
                }

                override fun onResponse(call: Call, response: Response) {
                    try {
                        val responseBody = response.body?.string()

                        if (response.isSuccessful) {
                            callback.onSuccess(responseBody ?: "")
                        } else {
                            callback.onError(responseBody ?: "")
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error processing response", e)
                        callback.onError(e.localizedMessage ?: "Error processing response")
                    }
                }
            })
        }
    }
}